### DISCUSSION TITLE
Nextjs App facing random timeout errors with turso

### DISCUSSION THREAD

**Discussion by subodhpareek18** [vercel/next.js#68269] [Help]
### Summary

It started a few days ago where out of no where my app started showing a client side except error, looking at the vercel logs ti showed the functions were timing out after 10s. After trying out a bunch of thing and failing I just deleted and re-setup the project on vercel which seemed to fix the issue. I thought this was the devops equivalent of stopping and starting the computer. 

Today the issue has come again, the same trick didn't work. So I though to increase the max function duration to 60s. And they are still timing out. Which makes no sense. 

Everything works perfectly fine on local, I'm using trpc with drizzle and turso. The requests to the DB take a second or 2 max. I thought this had something to do with Suspense boundaries so I removed those also. Nothing seems to be working. Is anybody else facing the same issue, how should I go about solving it?

![image](https://github.com/user-attachments/assets/20f3e68d-f580-4d92-96a5-c89262d123db)
![image](https://github.com/user-attachments/assets/adfe277f-4fc6-4de1-b58a-44aa654a4cf4)

**Comment by icyJoseph** [comment]
Perhaps your code is hitting a dead branch? Like, a Promise that won't settle? Are all env vars up to date? Nothing has expired, or roles that could've changed?

**Reply by subodhpareek18** [reply]
All env vars are up to date. I also tried shifting from a SSRd page with server trpc to a client only data fetching. And the requests on the client are timing out. I thought this was related to the DB but the same instance of turso is working completely fine on my local machine.

**Reply by subodhpareek18** [reply]
What is strange is 2 out of the  9 routes are working consistently fine with the same SSR trpc setup

**Reply by icyJoseph** [reply]
Maybe it's time to add some logging, see how far in the request process it gets, before it "hangs" if that's the case.

**Reply by subodhpareek18** [reply]
I added some logs, sharing both the console on local and the vercel logs (which are latest first). It looks like the specific db call is definitely getting stuck on prod server. Because there are a lot of calls being made to fetch the user for authorization at multiple levels which are going fine. Something specific about this call on vercel production is an issue. The same DB query is working fine when running locally.

![Screenshot 2024-07-29 at 9 38 31 PM](https://github.com/user-attachments/assets/ea833db1-38df-45eb-b931-e8447f1ed01c)
![Screenshot 2024-07-29 at 9 38 44 PM](https://github.com/user-attachments/assets/155abcb6-e50d-42ac-8e9b-4658e3bbdb03)

**Reply by subodhpareek18** [reply]
This is what this DB call looks like. And many such calls for other routes are failing in the same way.
```ts
  getBookingChannels: adminProcedure.query(async ({ ctx }) => {
    console.debug("[admin-router] fetching channels", { server: ctx.server });
    const channels = await db.query.Channels.findMany({
      where: eq(Channels.type, "booking"),
      with: {
        booking: {
          columns: {
            id: true,
            status: true,
            details: true,
          },
        },
      },
    });
    console.debug("[admin-router] fetched channels", { server: ctx.server });
    return channels;
  }),
```

Here ctx.server just tells me if the call was made client side or server side (SSR)

**Reply by subodhpareek18** [reply]
It's also happening with these queries (not exhaustive)
```ts
  getTasks: adminProcedure.query(async () => {
    console.debug("[admin-router] fetching tasks");
    const tasks = await db.query.Tasks.findMany({
      orderBy: [desc(Tasks.createdAt)],
    });
    console.debug("[admin-router] fetched tasks");
    return tasks;
  }),
```
```ts
 getTradesmen: adminProcedure.query(async () => {
    console.debug("[admin-router] fetching tradesmen");
    const tradesmen = await db.query.Tradesmen.findMany({
      with: {
        org: {
          columns: {
            id: true,
            name: true,
          },
        },
      },
    });
    console.debug("[admin-router] fetched tradesmen");
    return tradesmen;
  }),
```
```ts
getTradesmenChannels: adminProcedure.query(async () => {
    console.debug("[admin-router] fetching tradesmen channels");
    const channels = await db.query.Channels.findMany({
      where: eq(Channels.type, "tradesmen"),
      with: {
        tradesman: true,
      },
    });
    console.debug("[admin-router] fetched tradesmen channels");
    return channels;
  }),
```
```ts
 getInvoices: adminProcedure.query(async () => {
    return db.query.Invoices.findMany({
      orderBy: [desc(Invoices.createdAt)],
    });
  }),
```


but not these, routes with them are totally fine
```ts
 getOrgs: adminProcedure.query(async () => {
    return db.query.Orgs.findMany();
  }),
```
```ts
 getConfig: adminProcedure.query(() => {
    return db.query.Config.findMany();
  }),
```

The clear difference here is the above two are very starightforward queries with no `with/orderBy` columns. Excluding `where` from it, since the bottom two queries are working fine.

```ts
    console.debug("[trpc] fetching user", { email });
    const user = await db.query.Users.findFirst({
      where: eq(Users.email, email),
    });
    console.debug("[trpc] fetched user", { email });
```
```ts
      console.debug("[auth middleware] authorization started");
      const isOnDashboard = nextUrl.pathname.startsWith("/app");
      const user = auth?.user?.email
        ? await db.query.Users.findFirst({
            where: eq(Users.email, auth.user.email),
          })
        : null;
      console.debug("[auth middleware] user", { email: auth?.user?.email });
```

**Reply by icyJoseph** [reply]
And this query, runs against the production database locally, without issue?

I wonder if there's either: a race condition, an assumption that's somehow not holding up for that request, or a non linear algorithm? Like a query execution, user data, are done in such a way that it just takes a lot of time to complete? For example a O(n^2) operation.

**Reply by subodhpareek18** [reply]
I've tried 3 different instances, my usual local turso instance, the prod instance, and recently branched prod instance. All work fine on local, none on production.

**Reply by subodhpareek18** [reply]
To validate by `with/orderBy` theory. I removed the `orderBy` from invoices and deployed it. But that didn't solve the issue. So it's not that I guess.

So now I'm wondering if this has got to do with the relationship between these tables? But if that's the case why work on local.

**Reply by icyJoseph** [reply]
Not to derail the conversation, but, compared to the CPU hosting your app, how much more powerful is your dev machine? You can see what CPU specs you have on your Vercel dashboard.

I hope this is not like the fetch issue we had back in the first days of Server Components, where fetch would just hang for requests over a certain size, and it would hang at the deserialization step.

Are the responses large?

Same node version locally and on the deployment?

I'm also not by my laptop, so at the moment ,I can only do rubber ducking help

**Reply by subodhpareek18** [reply]
That's definitely not a derailment. On local dev I have a M1 Pro macbook and using bun. On vercel just using the free hobby tier right now, am going to switch to vercel pro tomorrow, waiting for some clearance.

For some of the requests the responses are large yes, but not for all of them.

I'm going to investigate this further along these lines tomorrow, by reducing the size of the records on all views that are hanging. And also switching to pro tier before anything. Plus might be useful to stop using bun and use the same nodejs as what vercel does.

**Reply by subodhpareek18** [reply]
I removed all sorts of authorization, and added logging to drizzle. Had no impact. Bottom half is before I did that, top is after.
![image](https://github.com/user-attachments/assets/d76f3b29-1f43-495b-af80-01f1f008ca5e)


Now a query as simple as this is timing out
![Screenshot 2024-07-30 at 7 49 12 AM](https://github.com/user-attachments/assets/867e3c5a-d030-4352-b864-470ea9766c2f)

This on my local (with the same prod DB) takes <200ms to return some 22KB of data. Hard to think this by itself could cause a compute/memory issue. But I will still explore those areas.
![Screenshot 2024-07-30 at 7 49 24 AM](https://github.com/user-attachments/assets/6debf8f0-1744-4ac1-834c-b969e1aba3bc)

**Reply by icyJoseph** [reply]
Could it be that your DB server drops connections after some time? Is there a keep alive header on the responses from the DB server?

**Reply by subodhpareek18** [reply]
I was actually in a bit of a hurry and because of other issues I was facing with Turso (sqlite), like the column renaming or notNull/default values. I ended up starting a migration path to Neon (postgres). And other than the refactoring that comes with it, the above issue seems to have been solved for now.

**Reply by icyJoseph** [reply]
🤔 oh wow that's rather anti climatic, but good for you, I hope this doesn't happen again 👍

**Reply by subodhpareek18** [reply]
Hehe that's true. Thank you for taking the time to share your input :)

**Reply by VaggelisKa** [reply]
Huh, i'm actually facing similar issues. I have a Remix app with drizzle and turso which i hadn't worked on for some time and when creating some feature envs today im getting those timeouts only on there. My Prod site is still working as expected.

The strange thing is that the timeouts are happening only when using the query syntax by drizzle, normal `db.whatever.select()` is resolving fine on <1s time. Any chance you were also using the query syntax when coming upon these issues?

**Reply by subodhpareek18** [reply]
Yes @VaggelisKa  I was using the query syntax, you can look at the exact examples in the messages above.

**Reply by subodhpareek18** [reply]
But if the query syntax is the issue, then shifting from turso to neon might not be the solution I was hoping for. Though for now, the timeouts have gone away.
